examples/seismic/tti/operators.py

```diff
@@ -420,5 +420,162 @@ def ForwardOperator(model, source, receiver, space_order=4,
+    vz = TimeFunction(name='vz', grid=model.grid, staggered=stagg_z,
+                      save=source.nt if save else None,
+                      time_order=1, space_order=space_order)
+
+    if model.grid.dim == 3:
+        vy = TimeFunction(name='vy', grid=model.grid, staggered=stagg_y,
+                          save=source.nt if save else None,
+                          time_order=1, space_order=space_order)
+
+    pv = TimeFunction(name='pv', grid=model.grid,
+                      save=source.nt if save else None,
+                      time_order=1, space_order=space_order)
+
+    ph = TimeFunction(name='ph', grid=model.grid,
+                      save=source.nt if save else None,
+                      time_order=1, space_order=space_order)
+    # Stencils
+    phdx = staggered_diff(ph, dim=x, order=space_order, stagger=left,
+                          theta=theta, phi=phi)
+    u_vx = Eq(vx.forward, damp_loc * vx - damp_loc * s / rho * phdx)
+
+    pvdz = staggered_diff(pv, dim=z, order=space_order, stagger=left,
+                          theta=theta, phi=phi)
+    u_vz = Eq(vz.forward, damp_loc * vz - damp_loc * s / rho * pvdz)
+
+    dvx = staggered_diff(vx.forward, dim=x, order=space_order, stagger=right,
+                         theta=theta, phi=phi)
+    dvz = staggered_diff(vz.forward, dim=z, order=space_order, stagger=right,
+                         theta=theta, phi=phi)
+
+    u_vy = []
+    dvy = 0
+    if ndim == 3:
+        phdy = staggered_diff(ph, dim=y, order=space_order, stagger=left,
+                              theta=theta, phi=phi)
+        u_vy = [Eq(vy.forward, damp_loc * vy - damp_loc * s / rho * phdy)]
+        dvy = staggered_diff(vy.forward, dim=y, order=space_order, stagger=right,
+                             theta=theta, phi=phi)
+
+    pv_eq = Eq(pv.forward,
+               damp_loc * (pv - s * rho / m * (delta * (dvx + dvy) + dvz)))
+
+    ph_eq = Eq(ph.forward,
+               damp_loc * (ph - s * rho / m * (epsilon * (dvx + dvy) + delta * dvz)))
+
+    # Source symbol with input wavelet
+    src = PointSource(name='src', grid=model.grid, time_range=source.time_range,
+                      npoint=source.npoint)
+    rec = Receiver(name='rec', grid=model.grid, time_range=receiver.time_range,
+                   npoint=receiver.npoint)
+    src_term = src.inject(field=pv.forward, offset=model.nbpml,
+                          expr=src * rho * dt / m)
+    src_term += src.inject(field=ph.forward, offset=model.nbpml,
+                           expr=src * rho * dt / m)
+    # Data is sampled at receiver locations
+    rec_term = rec.interpolate(expr=pv + ph, offset=model.nbpml)
+
+    # Substitute spacing terms to reduce flops
+    op = Operator([u_vx, u_vz] + u_vy + rec_term + [pv_eq, ph_eq] + src_term,
+                  subs=model.spacing_map,
+                  dse='advanced', dle='advanced')
+
+    return op
+
+
+def staggered_diff(f, dim, order, stagger=centered, theta=0, phi=0):
```

Comment:
there must be a way of generalising this, and have it in Devito

Reply:
Yes gonna be moved in finite_difference